transports/onion/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+#![doc(html_logo_url = "https://libp2p.io/img/logo_small.png")]
+#![doc(html_favicon_url = "https://libp2p.io/img/favicon.png")]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+
+use std::pin::Pin;
+use std::sync::Arc;
+
+use address::{dangerous_extract_tor_address, safe_extract_tor_address};
+use arti_client::{TorAddrError, TorClient, TorClientBuilder};
+use futures::{future::BoxFuture, FutureExt};
+use libp2p_core::{transport::TransportError, Multiaddr, Transport};
+use tor_rtcompat::Runtime;
+
+mod address;
+mod provider;
+
+#[doc(inline)]
+pub use provider::OnionStream;
+
+#[derive(Debug, thiserror::Error)]
+pub enum OnionError {
+    #[error("error during address translation")]
+    AddrErr(#[from] TorAddrError),
+    #[error("error in arti")]
+    ArtiErr(#[from] arti_client::Error),
+}
+
+#[derive(Clone)]
+pub struct OnionClient<R: Runtime> {
+    // client is in an Arc, because wihtout it the Transport::Dial method can't be implemented,
+    // due to lifetime issues. With the, eventual, stabilization of static async traits this issue
+    // will be resolved.
+    client: Arc<TorClient<R>>,
+    pub conversion_mode: AddressConversion,
```

Comment:
Instead of making this public, what do you think about a `with_address_conversion` function? That would allow people to write:

```rust
let client = OnionClient::default().await?.with_address_conversion(AddressConversion::IpAndDns);
```

Reply:
sounds good and convenient, however I don't think it conflicts with making conversion mode public. There might be reasons why someone would like to change the address conversion mode later after construction.

Reply:
The `with_` function does allow for that though, right? The only thing that is not possible is reading the mode but given that it is static, I am not sure why one would need that.

Another though I just had is that we could embed it in the type signature too. That could allow client code to require a "safe" OnionTransport in a type-safe way. Might be overkill though!

Reply:
> The with_ function does allow for that though, right? The only thing that is not possible is reading the mode but given that it is static, I am not sure why one would need that.

For sure, I just considered it less natural. I will however remove the `pub` modifier if thats whats necessary.

> Another though I just had is that we could embed it in the type signature too. That could allow client code to require a "safe" OnionTransport in a type-safe way. Might be overkill though!
It might be possible, but I think it's not necessary. The API is pretty obfuscated now anyway and I don't see a real benefit.

Reply:
I think this is a good compromise. Having the field pub allows users to read it and the `with` function makes it possible to chain it onto an existing instance.